crates/ty_server/src/server/api/semantic_tokens.rs

```diff
@@ -0,0 +1,61 @@
+use lsp_types::SemanticToken;
+use ruff_db::source::{line_index, source_text};
+use ruff_text_size::{Ranged, TextLen, TextRange};
+use ty_ide::semantic_tokens;
+use ty_project::ProjectDatabase;
+
+use crate::document::{PositionEncoding, TextSizeExt};
+
+/// Common logic for generating semantic tokens, either for full document or a specific range.
+/// If no range is provided, the entire file is processed.
+pub(crate) fn generate_semantic_tokens(
+    db: &ProjectDatabase,
+    file: ruff_db::files::File,
+    range: Option<TextRange>,
+    encoding: PositionEncoding,
+) -> Vec<SemanticToken> {
+    let source = source_text(db, file);
+    let line_index = line_index(db, file);
+
+    let requested_range = range.unwrap_or_else(|| TextRange::new(0.into(), source.text_len()));
+    let semantic_token_data = semantic_tokens(db, file, Some(requested_range));
+
+    // Convert semantic tokens to LSP format
+    let mut lsp_tokens = Vec::new();
+    let mut prev_line = 0u32;
+    let mut prev_start = 0u32;
+
+    for token in &*semantic_token_data {
+        let start_position = token.start().to_position(&source, &line_index, encoding);
+        let line = start_position.line;
+        let character = start_position.character;
+        let length = token.range().len().to_u32();
```

Comment:
I think the `length` here must respect the `encoding`. The value you get here is the byte length of the token, but that's only correct if the negogiated encoding is UTF8. 

Is there a reason why you can't use `text_range.to_lsp_range` instead?